service/awsconfig/v8/resource/lifecycle/current.go

```diff
@@ -95,34 +101,212 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
+		LifecycleHookName:     aws.String(key.NodeDrainerLifecycleHookName),
+	}
+
+	_, err := r.aws.AutoScaling.CompleteLifecycleAction(i)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("completed lifecycle hook action for guest cluster instance '%s'", instanceID))
+
+	return nil
+}
+
+func (r *Resource) createNodeConfig(ctx context.Context, customObject providerv1alpha1.AWSConfig, instanceID, privateDNS string) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "creating node config for guest cluster node")
+
+	n := v1.NamespaceDefault
+	c := &v1alpha1.NodeConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				key.InstanceIDAnnotation: instanceID,
+			},
+			Labels: map[string]string{
+				key.ClusterIDLabel: key.ClusterID(customObject),
+			},
+			Name: privateDNS,
+		},
+		Spec: v1alpha1.NodeConfigSpec{
+			Guest: v1alpha1.NodeConfigSpecGuest{
+				Cluster: v1alpha1.NodeConfigSpecGuestCluster{
+					API: v1alpha1.NodeConfigSpecGuestClusterAPI{
+						Endpoint: key.ClusterAPIEndpoint(customObject),
+					},
+					ID: key.ClusterID(customObject),
+				},
+				Node: v1alpha1.NodeConfigSpecGuestNode{
+					Name: privateDNS,
+				},
+			},
+			VersionBundle: v1alpha1.NodeConfigSpecVersionBundle{
+				Version: "0.1.0",
+			},
+		},
+	}
+
+	_, err := r.g8sClient.CoreV1alpha1().NodeConfigs(n).Create(c)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "created node config for guest cluster node")
+
+	return nil
+}
+
+func (r *Resource) deleteNodeConfig(ctx context.Context, privateDNS string) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleting node config for guest cluster node")
+
+	n := v1.NamespaceDefault
+	i := privateDNS
+	o := &metav1.DeleteOptions{}
+
+	err := r.g8sClient.CoreV1alpha1().NodeConfigs(n).Delete(i, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted node config for guest cluster node")
+
+	return nil
+}
+
+func (r *Resource) privateDNSForInstance(instanceID string) (string, error) {
+	i := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(instanceID),
+		},
+	}
+
+	o, err := r.aws.EC2.DescribeInstances(i)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	if len(o.Reservations) != 1 {
+		return "", microerror.Maskf(executionFailedError, "expected 1 reservation, got %d", len(o.Reservations))
+	}
+	if len(o.Reservations[0].Instances) != 1 {
+		return "", microerror.Maskf(executionFailedError, "expected 1 instance, got %d", len(o.Reservations[0].Instances))
+	}
+
+	privateDNS := *o.Reservations[0].Instances[0].PrivateDnsName
+
+	return privateDNS, nil
+}
+
+func hasFinalStatus(conditions []v1alpha1.NodeConfigStatusCondition) bool {
+	for _, c := range conditions {
+		if c.Type == "Drained" && c.Status == "True" {
+			return true
+		}
```

Comment:
See the node-operator updating to this state when draining is done: https://github.com/giantswarm/node-operator/blob/3a5fa43c1580928a137de99e2e9cfba50f65f3db/service/nodeconfig/v1/resource/node/create.go#L110-L113.

Reply:
For sure. Right now I would not do it because I do not expect this to change heavily or being used somewhere else. When we start to integrate draining into KVM we can add another library. Let me know if you want to get this sorted now though.